### ISSUE TITLE
ENH: stats.qmc.geometric_discrepancy: `method='mindist'` can improve memory complexity

### ISSUE THREAD

**Issue by MLopez-Ibanez** [scipy/scipy#22837]
### Is your feature request related to a problem? Please describe.

`geometric_discrepancy(method='mindist')` calculates a full pairwise distance matrix just to return the minimum value of the matrix. This requires quadratic memory, which has very poor scalability, as the example below shows.  With 32768 points, `geometric_discrepancy(method='mindist')` requires more than 4GB of memory!

### Describe the solution you'd like.

Chunking the computation per row as in the code below has the same quadratic computation time but only uses linear memory. 

```python
import numpy as np
import pandas as pd
import timeit
from scipy.stats import qmc
from scipy.spatial import distance
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme(style="darkgrid")

def pmindist(x):
    n = x.shape[0]
    d = np.inf
    for i in range(1, n):
        d = min(d, distance.cdist(x[i:, :], x[(i-1):i, :]).min())
    return d

res = []
for k in range(10,29):
    n = int(np.sqrt(2**k))
    x = np.random.default_rng().random((n, 2))
    start = timeit.default_timer()
    v1 = pmindist(x)
    t1 = timeit.default_timer()
    v2 = qmc.geometric_discrepancy(x)
    t2 = timeit.default_timer()
    assert v1 == v2
    res += [ dict(n=n, time=t1 - start, method="pmindist")]
    res += [ dict(n=n, time=t2 - t1, method="qmc")]
    
df = pd.DataFrame(res)
plt.figure() 
ax = sns.lineplot(data=df, x="n", y="time", hue="method", marker='o')
ax.set(yscale='log', xscale='log')
#plt.xticks(df['n'].unique(), [ fr'$2^{{{int(p)}}}$' for p in df['n'].unique()])
plt.show()
```

A C/C++ implementation would be even faster.

### Describe alternatives you've considered.

Scikit-learn has [`pairwise_distances_chunked`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances_chunked.html), which may be faster but more complex to implement.

### Additional context (e.g. screenshots, GIFs)

![Image](https://github.com/user-attachments/assets/cb910c21-7ea3-4fbc-802f-1b3c2ba9adbe)

**Comment by tylerjereddy** [comment]
With a little creativity I think we can achieve equivalent asymptotic space complexity and much better time complexity using a KDTree approach. Sample diff and benchmarks below. As an added advantage, the code is already in C++ under the hood and even accepts multiple workers (parallelism).

<details>
<summary> sample diff</summary>

```diff
--- a/probe.py
+++ b/probe.py
@@ -2,7 +2,7 @@ import numpy as np
 import pandas as pd
 import timeit
 from scipy.stats import qmc
-from scipy.spatial import distance
+from scipy.spatial import distance, KDTree
 import matplotlib.pyplot as plt
 import seaborn as sns
 sns.set_theme(style="darkgrid")
@@ -10,8 +10,12 @@ sns.set_theme(style="darkgrid")
 def pmindist(x):
     n = x.shape[0]
     d = np.inf
-    for i in range(1, n):
-        d = min(d, distance.cdist(x[i:, :], x[(i-1):i, :]).min())
+    tree = KDTree(x)
+    d, i = tree.query(x,
+                      k=2,
+                      workers=8,
+                      distance_upper_bound=distance.euclidean(x[..., 0], x[..., 1]))
+    d = d[..., 1].min()
     return d
 
 res = []
@@ -23,7 +27,7 @@ for k in range(10,29):
     t1 = timeit.default_timer()
     v2 = qmc.geometric_discrepancy(x)
     t2 = timeit.default_timer()
-    assert v1 == v2
+    assert np.allclose(v1, v2)
     res += [ dict(n=n, time=t1 - start, method="pmindist")]
     res += [ dict(n=n, time=t2 - t1, method="qmc")]
```

</details>

Sample results (obviously, a user would have to select core count based on problem size):

Single core:
![Image](https://github.com/user-attachments/assets/05c21e86-d318-4618-a63c-60043ab76e35)

Two cores:
![Image](https://github.com/user-attachments/assets/2c6a39b2-d6cb-4d4d-8900-47cf936a9a31)

Four cores:
![Image](https://github.com/user-attachments/assets/e02f39e3-d7ea-48b7-8879-cbeec7c162bf)

Eight cores:
![Image](https://github.com/user-attachments/assets/9e6f6168-b29d-4357-8174-37c4c9ec2dd2)

12 cores (!):
![Image](https://github.com/user-attachments/assets/64b5c259-134c-43eb-85c1-2310cf15549d)

Maybe we should check with @tupui before getting too excited though, since this is a `qmc` thing.

**Comment by tupui** [comment]
Oh nice findings 👍 by all mean, if you manage to get the same min dist and that works in higher dimensions too, let's go 😃👏

**Comment by MLopez-Ibanez** [comment]
A more detailed version that can be used within `stats.qmc.geometric_discrepancy`. 

```python
def pmindist_new(x, metric='euclidean', workers = 1):
    if metric == 'euclidean':
        p = 2
        distance_fun = distance.euclidean
    elif metric == 'cityblock':
        p = 1
        distance_fun = distance.cityblock
    elif metric == 'chebyshev':
        p = np.inf
        distance_fun = distance.chebyshev
    else:
        # Slow path for metrics unsupported by KDTree.
        distances = distance.pdist(sample, metric=metric)  # type: ignore[call-overload]
        return distances.min()

    distance_upper_bound = distance_fun(x[0,:], x[1,:])
    tree = KDTree(x)
    d, _ = tree.query(x,
                      k=[2], p = p,
                      workers=workers,
                      distance_upper_bound=distance_upper_bound)
    return d.min()
```

It would be nice if `workers` was added to `stats.qmc.geometric_discrepancy` so it can be passed to this function.

**Comment by MLopez-Ibanez** [comment]
@tylerjereddy had the idea of using KDTree. I don't want to take the credit from him, but I am happy for him to take my suggestions on how to implement this in  `stats.qmc.geometric_discrepancy`.

**Comment by tylerjereddy** [comment]
@MLopez-Ibanez feel free to open a PR, if you're a first time contributor it is probably quite valuable for us to get you engaged. I can always help Pamphile with the review and be involved that way.
